service/src/main/java/bio/terra/cbas/models/Run.java

```diff
@@ -14,4 +15,20 @@ public record Run(
   public UUID getRunSetId() {
     return runSet.id();
   }
+
+  public Run withStatus(String newStatus) {
+    return new Run(id, engineId, runSet, entityId, submissionTimestamp, newStatus);
+  }
+
+  public boolean isTerminal() {
+    RunState state = RunState.fromValue(this.status());
+    return RunState.CANCELED.equals(state)
+        || RunState.COMPLETE.equals(state)
+        || RunState.EXECUTOR_ERROR.equals(state)
+        || RunState.SYSTEM_ERROR.equals(state);
```

Comment:
If `RunState` is an enum, and `state` is non-`null`, I think it's more idiomatic to compare enums using `==` instead of `equals()`. Java enums are singleton objects to `==` is an OK way to compare them.

Or maybe a better way overall is to use an `EnumSet<>` to represent the known terminal states:

```
  private static final Set<RunState> TERMINAL_STATES = EnumSet.of(RunState.CANCELED, RunState.COMPLETE,
        RunState.EXECUTOR_ERROR, RunState.SYSTEM_ERROR);
  ...
  public boolean isTerminal() {
    return TERMINAL_STATES.contains(RunState.fromValue(status));
  }
```
This doesn't mind `null` (it returns `false`) but if `null` is possible maybe you'd want to handle that explicitly.

Reply:
Oh, I like the set contains a lot. 

FWIW, I'm curious what your take on this thought is: I think long term this mixing of WES and CBAS API enums within business logic (especially writing to and reading from the database) is going to cause headaches. Is it normal to want to create yet-another-almost-parallel enum for business logic and database usages, with converters into and out of potentially different sets of API enums only on the interfaces?

Reply:
I've thought about this but don't have a great answer. The main problem I see is with auto generated code like `RunState`. For this code, I think what you really want is to make `isTerminal()` a property of `RunState` directly but there's no easy way that I know of to add code to an openapi generated enum.

Using parallel objects, and only using the openapi generated classes in the controller layer, seems like a good approach. At the controller layer you convert to your "real" objects, and convert "real" objects back to openapi objects. So any code outside the controller doesn't need to know about these objects. You can make this conversion easier by adding code to your "real" objects that convert from/to the generated objects as necessary. For example: https://github.com/DataBiosphere/terra-data-catalog/blob/main/service/src/main/java/bio/terra/catalog/common/StorageSystem.java

I like this approach since it isolates logic from the openapi layer. But it's a lot of work to do for all your input/output objects, and even in a very small service like catalog we haven't been consistent in doing it.

If you can use the same object for both database and business logic then it seems worth it to create a parallel object.